File: php/tests/CycleTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use Oojs\Registry;
use PHPUnit\Framework\TestCase;

use function Oojs\validate;

/**
 * Tests for cycles in the object model (§8.13).
 *
 * OOJS explicitly allows cycles at both the schema level (type-reference chains
 * that loop back) and the instance level (finitely-nested standalone instances and
 * graph documents with $ref-id references that form cycles).
 *
 * Schema-level cycles are safe because TypeRef values are stored as plain strings —
 * no recursive expansion occurs during schema loading.
 *
 * Instance-level cycles in standalone instances cannot occur in pure JSON (which is
 * always a tree), so the validator terminates automatically.
 *
 * Cycles in graph documents are handled by type-checking $ref-id references without
 * recursive inline validation (§8.12, §8.13.2).
 */
class CycleTest extends TestCase
{
    // ------------------------------------------------------------------
    // §8.13.1 — Schema-level cycles
    // ------------------------------------------------------------------

    public function test_self_referential_type_schema_loads_without_error(): void
    {
        // Employee.manager is a TypeRef to Employee (self-referential).
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/self-ref',
            'types' => [
                'Employee' => [
                    'properties' => [
                        'employeeId' => ['type' => 'string'],
                        'name'       => ['type' => 'string'],
                        'manager'    => ['type' => 'Employee'],  // self-referential
                    ],
                    'required' => ['employeeId', 'name'],
                ],
            ],
        ]);
        $this->assertArrayHasKey('Employee', $schema->types);
    }

    public function test_two_type_cycle_schema_loads_without_error(): void
    {
        // Employee.department → Department, Department.head → Employee.
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/two-cycle',
            'types' => [
                'Employee' => [
                    'properties' => [
                        'employeeId' => ['type' => 'string'],
                        'name'       => ['type' => 'string'],
                        'department' => ['type' => 'Department'],
                    ],
                    'required' => ['employeeId', 'name'],
                ],
                'Department' => [
                    'properties' => [
                        'departmentId' => ['type' => 'string'],
                        'name'         => ['type' => 'string'],
                        'head'         => ['type' => 'Employee'],
                    ],
                    'required' => ['departmentId', 'name'],
                ],
            ],
        ]);
        $this->assertArrayHasKey('Employee',   $schema->types);
        $this->assertArrayHasKey('Department', $schema->types);
    }

    public function test_three_type_cycle_schema_loads_without_error(): void
    {
        // A.b → B, B.c → C, C.a → A.
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/three-cycle',
            'types' => [
                'A' => [
                    'properties' => ['id' => ['type' => 'string'], 'b' => ['type' => 'B']],
                    'required'   => ['id'],
                ],
                'B' => [
                    'properties' => ['id' => ['type' => 'string'], 'c' => ['type' => 'C']],
                    'required'   => ['id'],
                ],
                'C' => [
                    'properties' => ['id' => ['type' => 'string'], 'a' => ['type' => 'A']],
                    'required'   => ['id'],
                ],
            ],
        ]);
        $this->assertArrayHasKey('A', $schema->types);
        $this->assertArrayHasKey('B', $schema->types);
        $this->assertArrayHasKey('C', $schema->types);
    }

    // ------------------------------------------------------------------
    // §8.13.3 — Standalone instances: JSON is a tree, so cycles are impossible
    // ------------------------------------------------------------------

    public function test_self_referential_type_instance_no_manager_valid(): void
    {
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/self-ref2',
            'types' => [
                'Employee' => [
                    'properties' => [
                        'employeeId' => ['type' => 'string'],
                        'name'       => ['type' => 'string'],
                        'manager'    => ['type' => 'Employee'],  // optional — no manager
                    ],
                    'required' => ['employeeId', 'name'],
                ],
            ],
        ]);
        $emp = [
            '_type'      => 'Employee',
            'employeeId' => 'E-001',
            'name'       => 'Alice',
            // no 'manager' — optional field absent
        ];
        $this->assertSame([], validate($emp, $schema->types['Employee'], $schema, $r));
    }

    public function test_self_referential_type_instance_depth_three_valid(): void
    {
        // Alice manages Bob who manages Carol. Depth = 3, no infinite loop.
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/self-ref3',
            'types' => [
                'Employee' => [
                    'properties' => [
                        'employeeId' => ['type' => 'string'],
                        'name'       => ['type' => 'string'],
                        'manager'    => ['type' => 'Employee'],
                    ],
                    'required' => ['employeeId', 'name'],
                ],
            ],
        ]);
        $alice = [
            '_type'      => 'Employee',
            'employeeId' => 'E-001',
            'name'       => 'Alice',
            'manager'    => [
                '_type'      => 'Employee',
                'employeeId' => 'E-002',
                'name'       => 'Bob',
                'manager'    => [
                    '_type'      => 'Employee',
                    'employeeId' => 'E-003',
                    'name'       => 'Carol',
                    // Carol has no manager — recursion terminates here
                ],
            ],
        ];
        $this->assertSame([], validate($alice, $schema->types['Employee'], $schema, $r));
    }

    public function test_three_type_cycle_instance_valid(): void
    {
        // A embeds B embeds C; C has no further nesting.
        // The class schema has a cycle (C.a → A) but the instance is a finite tree.
        $r = new Registry();
        $schema = $r->loadDict([
            '$oojs' => '1.0',
            '$id'   => 'https://example.org/schemas/three-cycle-inst',
            'types' => [
                'A' => [
                    'properties' => ['id' => ['type' => 'string'], 'b' => ['type' => 'B']],
                    'required'   => ['id'],
                ],
                'B' => [
                    'properties' => ['id' => ['type' => 'string'], 'c' => ['type' => 'C']],
                    'required'   => ['id'],
                ],
                'C' => [
                    'properties' => ['id' => ['type' => 'string'], 'a' => ['type' => 'A']],
                    'required'   => ['id'],
                ],
            ],
        ]);
        $instance = [
            '_type' => 'A',
            'id'    => 'a-1',
            'b'     => [
                '_type' => 'B',
                'id'    => 'b-1',
                'c'     => [
                    '_type' => 'C',
                    'id'    => 'c-1',
                    // 'a' absent — optional, recursion terminates
                ],
            ],
        ];
        $this->assertSame([], validate($instance, $schema->types['A'], $schema, $r));
    }
}

File: java/src/test/java/org/oojs/CycleTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class CycleTest extends TestHelpers {

    @Test
    void self_referential_type_schema_loads_without_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/self-ref",
                "types", map(
                        "Employee", map(
                                "properties", map(
                                        "employeeId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "manager", map("type", "Employee")),
                                "required", list("employeeId", "name")))));

        assertTrue(schema.types.containsKey("Employee"));
    }

    @Test
    void two_type_cycle_schema_loads_without_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/two-cycle",
                "types", map(
                        "Employee", map(
                                "properties", map(
                                        "employeeId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "department", map("type", "Department")),
                                "required", list("employeeId", "name")),
                        "Department", map(
                                "properties", map(
                                        "departmentId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "head", map("type", "Employee")),
                                "required", list("departmentId", "name")))));

        assertTrue(schema.types.containsKey("Employee"));
        assertTrue(schema.types.containsKey("Department"));
    }

    @Test
    void three_type_cycle_schema_loads_without_error() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/three-cycle",
                "types", map(
                        "A", map("properties", map("id", map("type", "string"), "b", map("type", "B")), "required", list("id")),
                        "B", map("properties", map("id", map("type", "string"), "c", map("type", "C")), "required", list("id")),
                        "C", map("properties", map("id", map("type", "string"), "a", map("type", "A")), "required", list("id")))));

        assertTrue(schema.types.containsKey("A"));
        assertTrue(schema.types.containsKey("B"));
        assertTrue(schema.types.containsKey("C"));
    }

    @Test
    void self_referential_type_instance_no_manager_valid() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/self-ref2",
                "types", map(
                        "Employee", map(
                                "properties", map(
                                        "employeeId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "manager", map("type", "Employee")),
                                "required", list("employeeId", "name")))));

        Map<String, Object> employee = map(
                "_type", "Employee",
                "employeeId", "E-001",
                "name", "Alice");

        List<ValidationError> errs = new Validator(r).validate(employee, schema.types.get("Employee"), schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void self_referential_type_instance_depth_three_valid() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/self-ref3",
                "types", map(
                        "Employee", map(
                                "properties", map(
                                        "employeeId", map("type", "string"),
                                        "name", map("type", "string"),
                                        "manager", map("type", "Employee")),
                                "required", list("employeeId", "name")))));

        Map<String, Object> alice = map(
                "_type", "Employee",
                "employeeId", "E-001",
                "name", "Alice",
                "manager", map(
                        "_type", "Employee",
                        "employeeId", "E-002",
                        "name", "Bob",
                        "manager", map(
                                "_type", "Employee",
                                "employeeId", "E-003",
                                "name", "Carol")));

        List<ValidationError> errs = new Validator(r).validate(alice, schema.types.get("Employee"), schema);
        assertEquals(List.of(), errs);
    }

    @Test
    void three_type_cycle_instance_valid() {
        Registry r = new Registry();
        Schema schema = r.loadMap(map(
                "$oojs", "1.0",
                "$id", "https://example.org/schemas/three-cycle-inst",
                "types", map(
                        "A", map("properties", map("id", map("type", "string"), "b", map("type", "B")), "required", list("id")),
                        "B", map("properties", map("id", map("type", "string"), "c", map("type", "C")), "required", list("id")),
                        "C", map("properties", map("id", map("type", "string"), "a", map("type", "A")), "required", list("id")))));

        Map<String, Object> instance = map(
                "_type", "A",
                "id", "a-1",
                "b", map(
                        "_type", "B",
                        "id", "b-1",
                        "c", map("_type", "C", "id", "c-1")));

        List<ValidationError> errs = new Validator(r).validate(instance, schema.types.get("A"), schema);
        assertEquals(List.of(), errs);
    }
}
